lib/client/mfa.go

```diff
@@ -18,72 +18,75 @@ package client
 
 import (
 	"context"
-	"errors"
 	"fmt"
-	"io"
 	"os"
 	"strings"
 	"sync"
+	"sync/atomic"
 
 	"github.com/gravitational/teleport/api/client/proto"
-	"github.com/gravitational/teleport/lib/utils/prompt"
 	"github.com/gravitational/trace"
 
 	wanlib "github.com/gravitational/teleport/lib/auth/webauthn"
 	wancli "github.com/gravitational/teleport/lib/auth/webauthncli"
 )
 
-type (
-	OTPPrompt func(ctx context.Context, out io.Writer, in prompt.Reader, question string) (string, error)
-	WebPrompt func(
-		ctx context.Context,
-		origin, user string,
-		assertion *wanlib.CredentialAssertion,
-		prompt wancli.LoginPrompt) (*proto.MFAAuthenticateResponse, string, error)
-)
+// promptWebauthn provides indirection for tests.
+var promptWebauthn = wancli.Login
 
-// PlatformPrompt groups functions that prompt the user for inputs.
-// It's purpose is to allow tests to replace actual user prompts with other
-// functions.
-type PlatformPrompt struct {
-	// OTP is the OTP prompt function.
-	OTP OTPPrompt
-	// Webauthn is the WebAuth prompt function.
-	Webauthn WebPrompt
+type stdinRead struct {
+	value string
+	err   error
 }
 
-func (pp *PlatformPrompt) Reset() *PlatformPrompt {
-	pp.Swap(prompt.Input, wancli.Login)
-	return pp
+// stdinHijack hijacks stdin for a single password-like read.
+// After startRead is called the read will be sent to the C channel.
+type stdinHijack struct {
+	C       chan stdinRead
+	started int32
 }
 
-func (pp *PlatformPrompt) Swap(otp OTPPrompt, web WebPrompt) {
-	pp.OTP = otp
-	pp.Webauthn = web
+func (h *stdinHijack) startRead() {
+	if !atomic.CompareAndSwapInt32(&h.started, 0, 1) {
```

Comment:
Is there a case where we expect more than one `startRead` but exactly one read from `stdinHijack.C` channel? 
If we do something like code below twice, someone will hang:
```go
stdin.startRead()
text:= <-stdin.C
```

Reply:
Yes, there is scenario that works like this:

1. The user has both OTP and WebAuthn enabled (hence gets the "combined" prompt)
2. The user chooses webauthn by tapping a security key
3. Their security key asks for a PIN

In the scenario above, the OTP password read has to be "redirected" to the PIN read. The user would have to be incredibly unlucky to tap both the security key and enter at a moment that gets it to hang. I've added a test to cover this and some concurrency control, but bear in mind that we can't both eliminate the race and keep the current UX.